src/lib/core/style/add-focus-classes.ts

```diff
@@ -0,0 +1,80 @@
+import {Directive, Injectable, Optional, SkipSelf} from '@angular/core';
+
+
+/** Singleton that allows all instances of CdkAddFocusClasses to share document event listeners. */
+@Injectable()
+export class CdkFocusCauseDetector {
+  /** Whether a keydown event has just occurred. */
+  get keydownOccurred() { return this._keydownOccurred; }
+  private _keydownOccurred = false;
+
+  get mousedownOccurred() { return this._mousedownOccurred; }
+  private _mousedownOccurred = false;
+
+  constructor() {
+    document.addEventListener('keydown', () => {
+      this._keydownOccurred = true;
+      setTimeout(() => this._keydownOccurred = false, 0);
```

Comment:
I'm slightly concerned about the performance implications of scheduling a asynchronous action on every keydown in the application. What can we do to conclusively say the cost for this would be minor?